##=== STARTUP ===##

#load libraries-----
library(dplyr, ggplot2, data.table, tidyr)

#load dataframes------
df <- read.table("geneouts", header=T) #dataframe generated by SNPGenie
id <- read.table("id.txt", header=T) #file with file names and whether they're feral or managed
imm_genes <- read.table("imm_genes", header=T, fill=T) #file with GB IDs, gene function, and gene pathway
gb_conv <- read.table("GB_GBold", header=T, fill=T) #table of corresponding new and old GB names
my.ncbi <- read.table("GB_to_NCBI", fill=T) #table of conversions from NCBI rna transcripts to GB
	colnames(my.ncbi) <- c("product", "GBold")

#dataframe cleanup-------
df <- subset(df, select=-c(1)) #get rid of combined column 1
df$product <- gsub('rna-', '', df$product) #gets rid of rna- in names for easier later combinations
df$NS <- df$piN/df$piS #creates piN/piS column
df <- na.omit(df) #gets rid of NaNs/NAs
df <- df[!is.infinite(df$NS),] #gets rid of Infs
merge(id, merge(my.ncbi, df, by="product"))
#written table = GenDiv

##== TABLE OF AVERAGE AND STANDARD DEVIATION ==##

df <- read.table("GenDiv", header=T)

#Avg. NS by sample--------
table.means <- aggregate(df$NS~df$Field2, df, mean)
table.sd <- aggregate(df$NS~df$Field2, df, sd)
table.ns.sd <- cbind(table.means, table.sd)
  table.ns.sd <- subset(table.ns.sd, select=-c(3))
  colnames(table.ns.sd) <- c("sample", "mean_ns", "sd")
  write.table(table.ns.sd, "stock.ns", sep="\t", col.names=T, row.names=F)

table.means <- aggregate(df$piS~df$Field2, df, mean)
table.sd <- aggregate(df$piS~df$Field2, df, sd)
table.ns.sd <- cbind(table.means, table.sd)
  table.ns.sd <- subset(table.ns.sd, select=-c(3))
  colnames(table.ns.sd) <- c("sample", "mean_piS", "sd")
  write.table(table.ns.sd, "stock.s", sep="\t", col.names=T, row.names=F)


#PLOT GENOME WIDE NS BY GROUP---------
u <- aggregate(df$NS, by=list(df$Field2), function(x) mean(x, na.rm=TRUE)) #create mean as point on ggplot
piS.plot <- ggplot(df, aes(x=Field2, y=NS, color=Field2)) + 
  geom_point(cex=1.5, pch=1.0, position = position_jitter(w=0.1, h=0)) +
  geom_point(data=u, cex=2, pch=19, aes(x=Group.1, y=x), colour="black") +
  theme_classic() + theme(legend.position = "none",
                          axis.title.x = element_text(color="black", size=10),
                          axis.text.y = element_text(color="black", size=10)) +
  ylab("piS") + xlab("Stock") + geom_hline(yintercept=c(0.00530899), lty=3) +
  ggtitle("piS by Stock")


#SIGNIFICANCE TESTING---------
aov <- aov(piS~Field2, data=df)
summary(aov)
tuk <- TukeyHSD(aov)
plot(tuk, las=2, col="red", cex.lab=0.5)

#PULL OUT HIGHEST % VALUE SNPs-----------
n <- 25 #for top 25%
top25 <- df[df$NS > quantile(df$NS,prob=1-n/100),]
write.table(top25, "top25_NS", sep="\t", col.names=T, row.names=F)

#CALCULATE SIGNIFICANTLY DIFFERENT GENES B/W FERAL AND MANAGED--------
df2 <- df[-grep("Ref", df$Field1),] #gets rid of ref SNPs in Field1 so the t-test is Managed/Feral
uniq.gene.id <- unique(df2$GBold)
gen.pval <- c()
gene.list <- c()
for(i in uniq.gene.id) {
  print(i)
  temp <- df2[which(df2$GBold==i),]
  length1 <- length(temp$Field1[which(temp$Field1=='Feral')])
  length2 <- length(temp$Field1[which(temp$Field1!='Feral')])
  length3 <- length(unique(temp$samp))
  if(length1>1 & length2>1 & length3>3) {
    temp2 <- t.test(temp$NS ~ temp$Field1)
    gen.pval <- c(temp2$p.value, gen.pval)
    gene.list <- c(i, gene.list)
  }
}
#ok so basically this file looks at all the gene IDs and does a t-test between each pair of genes between managed and feral, outputting two files called gene.list (the names of all the genes) and gen.pval (p-value of each t-test). also it prints (in console) the names of every gene it's working on so you can see exactly when it stalls out.

all.pval.ns.gene <- as.data.frame(cbind(gene.list, as.numeric(as.character(gen.pval))))
colnames(all.pval.ns.gene) <- c("gene", "pval") 
all.pval.ns.gene <- all.pval.ns.gene[!(all.pval.ns.gene$pval=="NaN"),]
all.pval.ns.gene$q <- qvalue1(all.pval.ns.gene2$pval)$q #but that's not all. also gotta calculate q-value to correct for false discovery rate because it's a lot of small prob calculations
plot(-log10(all.pval.ns.gene$gene),all.pval.ns.gene$q)
  abline(v=-log10(0.05))
  abline(h=0.05)

pval <- all.pval.ns.gene[(all.pval.ns.gene$q<=0.05),] #only if q is less than 0.05
write.table(pval, "gen_div_pval", col.names=T, row.names=F, quote=F)
imm <- left_join(imm_genes, pval, by=c("GB"="gene"))
imm <- na.omit(imm)

#if you need just one gene do this------
temp <- df2[which(df2$GBold=='GB44159'),] #DSCAM
temp2 <- t.test(temp$NS~temp$Field1)
temp2$p.value
p <- ggplot(temp, aes(x=Field1, y=NS, fill=Field1)) + geom_boxplot(outlier.color="black", notch=FALSE) + labs(x="Management", y="NS") + theme_classic() + theme(legend.position="none", axis.title.x=element_text(color="black", size=12), axis.text.y=element_text(color="black", size=12)) + ylab("𝜋N/𝜋S") + xlab("Management Practice") + ggtitle("Dscam")

#PULL OUT IMMUNE GENES------
imm_dt <- left_join(imm_genes, df, by=c("GB"="GBold"))
df2 <- left_join(imm_genes, df, by=c("GB"="GB"))
imm_df <- rbind(imm_dt, df2)
test <- imm_dt[complete.cases(imm_dt$samp),]

u <- aggregate(imm_dt$NS, by=list(imm_dt$Field2), function(x) mean(x, na.rm=TRUE))
piS.plot <- ggplot(imm_dt, aes(x=Field2, y=NS, color=Field2)) + 
  geom_point(cex=1.5, pch=1.0, position = position_jitter(w=0.1, h=0)) +
  geom_point(data=u, cex=2, pch=19, aes(x=Group.1, y=x), colour="black") +
  theme_classic() + theme(legend.position = "none",
                          axis.title.x = element_text(color="black", size=10),
                          axis.text.y = element_text(color="black", size=10)) +
  ylab("piS") + xlab("Stock") + geom_hline(yintercept=c(0.00530899), lty=3) +
  ggtitle("piS by Stock")