---
title: "amelpool_gendiv"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{amelpool_gendiv}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(amel.poolseq)
library(dplyr)
library(tidyr)
library(ggplot2)
```

# Genetic Diversity Tables and Figures
## From raw
```{r}
df <- hbpop_snpdf(GenDiv, popnames)
```

## From cluster
```{r}
df <- read.table("GenDiv", header=T)
gb_gbo <- read.table("GB_GBold", header=T)
gb_ncbi <- read.table("GB_to_NCBI", header=T, fill=T)

library(wrapr)

colnames(gb_ncbi) <- c("Transcript", "Gene")
gb_gbo$comb <- gb_gbo$GBold %?% gb_gbo$GB 
gb_ncbi_conv <- full_join(gb_gbo, gb_ncbi, join_by(comb==Gene), multiple="all") %>%
 select(!GBold) %>%
 distinct(Transcript, .keep_all=TRUE)

df <- df %>% select(!(c(GBold, GB))) %>% 
  left_join(gb_ncbi_conv, by=c("product" = "Transcript")) %>%
  select(!(c(GB)))

```

## Table 2: Avg. GenDiv by stock
```{r}
gd_summ <- df %>% 
  group_by(Field2) %>% 
  filter(Field2!="Clin") %>% filter(Field2!="MLin") %>% filter(Field2!="YLin") %>%
  summarize(
      avgS=mean(piS), sdS=sd(piS),
      avgNS=mean(NS), sdNS=sd(NS)
      )


## Pearsons correlation to mitochondrial diversity
```{r}
mito_div <- data.frame(stock=c("Cali", "KYFeral", "NCFeral", "PAFeral", "PAManaged", "PurdueMB", "Russ", "Sask", "TX"),
    hapdiv=as.numeric(c("0", "0.940", "0.950", "1.09", "1.03", "0", "0", "0", "0")))

gd_pear <- gd_summ %>% select(c("avgS")) %>% bind_cols(mito_div)
cor(gd_pear$hapdiv, gd_pear$avgS, method='pearson')

```

## Significance testing
```{r}
gd_plot <- df[-grep("Ref", df$Field1),]
aov <- aov(piS~Field2, data=gd_plot)
summary(aov)
tuk <- TukeyHSD(aov)
plot(tuk, las=2, col="red", cex.lab=0.5)
```

## SNPs exhibiting greatest directional selection
```{r}
n <- 25 #for top 25% of values
top25 <- df[df$NS > quantile(df$NS,prob=1-n/100),]
```

## Genes associated with immune function
```{r}
imm_dt <- left_join(imm_genes, df, by=c("GBID"="GBold"))
```

```{r echo=FALSE}
u <- aggregate(imm_dt$NS, by=list(imm_dt$Field2), function(x) mean(x, na.rm=TRUE))
piS.plot <- ggplot(imm_dt, aes(x=Field2, y=NS, color=Field2)) +
  geom_point(cex=1.5, pch=1.0, position = position_jitter(w=0.1, h=0)) +
  geom_point(data=u, cex=2, pch=19, aes(x=Group.1, y=x), colour="black") +
  theme_classic() + theme(legend.position = "none",
                          axis.title.x = element_text(color="black", size=10),
                          axis.text.y = element_text(color="black", size=10)) +
  ylab("piS") + xlab("Stock") + geom_hline(yintercept=c(0.00530899), lty=3) +
  ggtitle("piS by Stock")
```

# Significant differences in feral and managed pop genetic diversity
This loop looks at all the gene IDs and does a t-test between each pair of genes between managed and feral, outputting two files called gene.list (the names of all the genes) and gen.pval (p-value of each t-test). It prints (in console) the names of every gene it works on so you can see exactly when it stalls out.

```{r eval=FALSE}
df2 <- df[-grep("Ref", df$Field1),]
uniq.gene.id <- unique(df2$comb)
gen.pval <- c()
gene.list <- c()
for(i in uniq.gene.id) {
  #print(i)
  temp <- df2[which(df2$comb==i),]
  length1 <- length(temp$Field1[which(temp$Field1=='Feral')])
  length2 <- length(temp$Field1[which(temp$Field1!='Feral')])
  length3 <- length(unique(temp$samp))
  if(length1>1 & length2>1 & length3>3) {
    temp2 <- t.test(temp$NS ~ temp$Field1)
    gen.pval <- c(temp2$p.value, gen.pval)
    gene.list <- c(i, gene.list)
  }
}

all.pval.ns.gene <- data.frame(cbind(gene.list, as.numeric(as.character(gen.pval))))
colnames(all.pval.ns.gene) <- c("gene", "pval") 
pval_ns <- all.pval.ns.gene[!(all.pval.ns.gene$pval=="NaN"),]
pval_ns$pval <- as.numeric(pval_ns$pval)

```
## *q* value
*q*-value for each gene must also be calculated to correct for the false discovery rate because the previous loop executes many small probability calculations.

```{r}
pval_ns$q <- qvalue1(pval_ns$pval)$q 
plot(-log10(pval_ns$gene), pval_ns$q)
  abline(v=-log10(0.05))
  abline(h=0.05)

pval <- pval_ns[(pval_ns$q<=0.05),]
#write.table(pval, "gen_div_pval", col.names=T, row.names=F, quote=F)
```

## Pull out genes associated with immune function
```{r}
imm_genes <- data(imm_genes)
imm <- left_join(imm_genes, pval, by=c("GB"="gene"))
imm <- na.omit(imm)
```

## If using cluster
```{r}
brut <- read.table("Brutscher_Imm_Gene_ID", header=T, fill=T, row.names=NULL, sep="\t")
brut <- brut %>% left_join(gb_ncbi_conv, by=join_by(Gene.ID==Transcript)) %>% select(!GB)

brut2 <- brut[grep("GB", brut$Gene.ID),]
brut2 <- brut2 %>% 
  relocate(comb, .before=Gene.ID) %>% 
  rename(
    Gene.ID=3,
    comb=4) %>% 
  left_join(gb_ncbi_conv, by=join_by(comb==comb)) %>%
  select(!c(5,6))

brut_ev2 <- brut[-grep("GB", brut$Gene.ID),]
brut_ev3 <- rbind(brut_ev2, brut2)

harp <- read.table("imm_genes", header=T)
colnames(harp) <- c("GB", "Gene.Name", "Pathway")
harp$Gene.Name <- NA
harp$Pathway <- 'Hygiene'
harp2 <- harp %>% 
  left_join(gb_ncbi_conv, by=join_by(GB==comb), multiple="all") %>%
  select(!GB.y) %>%
  relocate(Gene.Name) %>%
  relocate(Pathway, .after=Gene.Name) %>%
  relocate(GB, .after=last_col()) %>%
  rename(
    Gene.ID=3,
    comb=4
    )

br_ha <- rbind(brut_ev3, harp2)
immgene_list <- br_ha[!duplicated(br_ev_ha$Gene.ID, incomparables=NA),]

colnames(immgene_list) <- c("Gene.Name", "Pathway", "Transcript", "GBID")

imm <- left_join(immgene_list, pval, by=c("GBID"="gene"))
imm <- na.omit(imm)

```

### Look at single genes
`hbpop_snpsing` allows you to look at differences in genetic diversity between specific genes. This example uses DSCAM.

```{r}
hbpop_snpsing(gd_plot, 'GB44159')
```


